### PR TITLE
(PUP-8356) Base fips specific behaviors without using facts

### PR DESCRIPTION
**Pull request by jsane**
We need to adjust some settings based on whether fips mode is enabled
or not for FIPS compliance. Till now the later was done using a fact (PUP-8141).
However that created a dependency in puppetserver which cannot load facts in certain configurations like dev testing and running from sources.
This change implements fips mode detection using a local method instead of facter.

### CODE REVIEW

File: lib/puppet/util/fips.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      false
+    end
+
+    # REMIND - How to check if we are on posix+linux system to avoid
+    # getting surprised on any odd-ball system that might not entertain
+    # the below file query.
+
+    fips_status_file = '/proc/sys/crypto/fips_enabled'
+
+    if File.exist?(fips_status_file) && File.open(fips_status_file, &:readline)[0].chr == '1'
```
Comment (joshcooper) [inline lib/puppet/util/fips.rb]: If the file is empty, I think this will raise an error due to the `[0]`. You might be able to use `Array#first` instead which should return the first line or `nil`

File: lib/puppet/util/fips.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+module Puppet::Util::FIPS
+  include Puppet::Util::Platform
+
+  def self.fips_enabled?
+    # We currently only plan to support FIPS on RHEL7
+    # so we try to eliminate anything un-supported using
+    # what we have available in lieu of not being able to specifically
+    # check for it.
+    if Puppet::Util::Platform.windows?
+      false
```
Comment (joshcooper) [inline lib/puppet/util/fips.rb]: I think you want `return false if Puppet::Util::Platform.windows?` otherwise we fall through and continue along.

File: lib/puppet/util/fips.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+require 'puppet/util/platform'
+# Method to determine whether where are running on a FIPs enabled host
```
Comment (joshcooper) [inline lib/puppet/util/fips.rb]: Since this is a platform-y thing, like `windows?`, let's just add the `fips_enabled?` method to `Puppet::Util::Platform`.

File: lib/puppet/defaults.rb [hunk 4]
```diff
@@ -1,4 +1,7 @@
+require 'puppet/util/fips'
+
 module Puppet
+  include Puppet::Util::FIPS
```
Comment (joshcooper) [inline lib/puppet/defaults.rb]: As long as you specify `module_function :fips_enabled?`, then you can take this include out, and just call it like you're doing below.

**Comment by jsane** [comment]
Pardon the delay; will take care of them today.

**Comment by jsane** [comment]
I now wonder if I need to condition the valid values of checksum types (default.rb ~line 951) to account for fips in addition to just supported_checksum_types. Right now we adjust those in some of the tests; may be they need to adjusted here itself.

File: lib/puppet/util/platform.rb [hunk 5]
```diff
@@ -17,6 +17,29 @@ def default_paths
...
+        # the below file query.
+
+        fips_status_file = '/proc/sys/crypto/fips_enabled'
+
+        if File.exist?(fips_status_file) && File.size(fips_status_file) > 1 &&
+           File.open(fips_status_file, &:readline)[0].chr == '1'
+          true
+        else
+          false
+        end
```
Comment (joshcooper) [inline lib/puppet/util/platform.rb]: The `fips_enabled` file has a 0-length, so this check is no good:

```
[root@wwxwbjmbdegguig ~]# ls -la /proc/sys/crypto/fips_enabled
-r--r--r-- 1 root root 0 Jan 22 03:15 /proc/sys/crypto/fips_enabled
```

I would simplify this by just trying to read at most 1 byte from the file.

```ruby
begin
  File.read(fips_status_file, 1) == '1'
rescue
  false
end
```

**Comment by joshcooper** [comment]
I think we should file a separate ticket for the case where the manifest says to use `md5` but the agent doesn't support it. Currently in that scenario the agent will segfault. Instead I would expect a graceful error that the checksum type is not supported by the agent.

```puppet
class fips {
  file { '/tmp/file.txt':
    ensure => file,
    source => 'puppet:///modules/fips/file.txt',
    checksum => 'md5'
  }
}
```

```
# puppet agent -t --http_debug
<- "GET /puppet/v3/file_metadata/modules/fips/file.txt?environment=production&links=manage&checksum_type=md5&source_permissions=ignore HTTP/1.1\r\nAccept: application/json, text/pson\r\nX-Puppet-Version: 5.3.4\r\nAccept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3\r\nUser-Agent: Puppet/5.3.4 Ruby/2.0.0-p648 (x86_64-linux)\r\nHost: mw0mycc5dfkjkwj.delivery.puppetlabs.net:8140\r\n\r\n"
-> "HTTP/1.1 200 OK\r\n"
-> "Date: Mon, 22 Jan 2018 18:59:43 GMT\r\n"
-> "Content-Type: application/json;charset=utf-8\r\n"
-> "X-Puppet-Version: 5.3.3\r\n"
-> "Vary: Accept-Encoding, User-Agent\r\n"
-> "Content-Encoding: gzip\r\n"
-> "Content-Length: 206\r\n"
-> "Server: Jetty(9.4.z-SNAPSHOT)\r\n"
-> "\r\n"
reading 206 bytes...
-> "\x1F\x8B\b\x00\x00\x00\x00\x00\x00\x005\x8FQ\x8E\xC3 \fD\xEF\xE2oT6\x94fK.\xB3\"\xE0mQ\t 0\xD9VQ\xEE\xBE\xA0\xB4?\xD6\x8C\xE47\x1Eo\x904\xDDa\x02\x8Edx\xAA)!y=\x17n\xA2E\x8Eau9\x86\x05\x03\x15\x9Er\xB4\xD5\x90\x8B\x81/My,\xFC\xD7\xA5>\x0E\xE9\xF1DO\x02\x06\x19\xBD&\xB7\xE2\xCF\x11\x1D\xAA\xF7\f\xBC\v\x8F\xD2\xEE,:\xE8\e\xB6\xAD\xF8\x170\xC3\xA4\x94bp\xCB\xB1\xA6\xB7n\xD9\b\x93\x14_\f\xCC\x1D\xCD\xA3\xD4\x05\xA6\r\xE8\x95\xB0\xE3\xF6\xD2\xD8U\xFB\xDA\xDD\xD6\xEC>\x0FJ\x8E\xDA(!\x95\xB0\xE2,\xBF\xCD(\xCE\x97YZ1\x0E\xC3p\x95\xB0\xB3\x0F\xDDK6\xDCb!\x17t\x7F\xE5\xA8\xB7\xFF\x033\xBB\xB7t\a\x01\x00\x00"
read 206 bytes
Conn keep-alive
md5_dgst.c(82): OpenSSL internal error, assertion failed: Digest MD5 forbidden in FIPS mode!
Aborted
```

Note it does work correctly if don't specify a checksum in the manifest, or you specify a non-md5 version.

**Comment by jsane** [comment]
Created PUP-8378 for intercepting md5 usages at runtime.

File: lib/puppet/util/platform.rb (hunk 5 above)
Comment (jsane) [reply lib/puppet/util/platform.rb]: rspec is tripping on the above use for ruby 1.9.3...have restarted it just to see if it is transient but based on the error does not look like. May need to go either with the older rev or something different.

File: lib/puppet/util/platform.rb [hunk 6]
```diff
@@ -17,6 +17,24 @@ def default_paths
...
+        # We currently only plan to support FIPS on RHEL7
+        # so we try to eliminate anything un-supported using what we have available
+        # in lieu of not being able to specifically check for it.
+
+        return false if Puppet::Util::Platform.windows?
+
+        fips_status_file = '/proc/sys/crypto/fips_enabled'
+
+        begin
+           File.open(fips_status_file, &:readline)[0].chr == '1'
```
Comment (joshcooper) [inline lib/puppet/util/platform.rb]: You can shorten this to just `File.read(fips_status_file, 1)`, which will only read and return only the first byte. Also recommend making `fips_status_file` a frozen constant.

**Comment by jsane** [comment]
thx. Reason I did not use Puppet::FileSystem.exist? is because it was getting into a circular dependency.
